### ISSUE TITLE
Multiple nodes in Docker-compose

### ISSUE THREAD

**Issue by andreaskring** [aldica/aldica#53]
Work on the feature supporting multiple repository and Share nodes in the `docker-compose.yml` is ongoing in the branch/folder: [https://github.com/aldica/aldica/tree/feature/multiple-nodes-docker-compose/docker](https://github.com/aldica/aldica/tree/feature/multiple-nodes-docker-compose/docker). Everything seems to be working except for an issue with Share (probably related to the configuration of load balancing). The following has been done:
* The Dockerfiles for Alfresco and Share have been updated
* Two Alfresco repo containers have been configured in `docker-compose.yml`
* Two Share containers have been configured in `docker-compose.yml`
* A Dockerfile for an Nginx load balancer has been added and the load balancer has also been added to the `docker-compose.yml`

**Things to note**
There were some problems with the Share CSRF filter and these were addressed as follows:
* A version of [share-config-custom.xml](https://github.com/aldica/aldica/blob/feature/multiple-nodes-docker-compose/docker/share-config-custom.xml) with the CSRF filter disabled was [attempted to be bind mounted ](https://github.com/aldica/aldica/blob/feature/multiple-nodes-docker-compose/docker/docker-compose.yml#L233) into the Share containers, but this was not possible due to the way the [upstream Share Dockerfile entrypoint is constructed](https://github.com/Alfresco/share/blob/6.0/packaging/docker/substituter.sh#L19), i.e. the `sed` commands in this file are not working due to permission issues when `share-config-custom.xml` is bind mounted into the container.
* The CSRF policies were attempted to to be set using the environment variables `CSRF_FILTER_REFERER` and `CSRF_FILTER_ORIGIN` (leaving these [empty](https://github.com/aldica/aldica/blob/feature/multiple-nodes-docker-compose/docker/docker-compose.yml#L218)), but this did not have the desired effect.
* Finally a (sub-optimal) solution was chosen, i.e. to [copy the `share-config-custom.xml` into the Docker images itself](https://github.com/aldica/aldica/blob/feature/multiple-nodes-docker-compose/docker/Dockerfile.share.download#L10). This works, but it require that we set the Share custom configurations directly in the Share Dockerfile.

With the CSRF configuations addressed as just described, everything seems to be working except for the load balancing for the Share containers. When starting the stack from the `docker` folder with
```
$ docker-compose up -d --build
```
and having set
```
127.0.0.1	localhost alfresco1 alfresco2 loadbalancer share1 share2
```
in the host `/etc/hosts`, it can be seen from the logs that the systems start without errors, and it can also be verified that the distributed nature of the caching mechanism for the repos is working (by doing the authentication ticket test). But when going to `http://loadbalancer:8080/share/page` and logging in, the login `POST` request is made (giving a HTTP status 302), but after the redirect, the user is just faced with the login page again...

@AFaust This is probably due to some Share configuration detail that I am missing - do you have an idea what could be wrong?

**Comment by AFaust** [comment]
Default Alfresco images' handling of CSRF is annoyingly trivial. One of many reasons I typically do not use Alfresco's images for anything other than community projects.

* For your sub-optimal solution you can also choose to include a share-config-custom-dev.xml file in the Docker image which only has to contain a CSRF config override with an effectively disabled state, and thus avoiding the override of all the other config.
* Using an empty value for origin / referer is not the same as effectively disabling CSRF. Without a configured value, there is an internal fallback using localhost / Tomcat server name for validation. In order to effectively disable CSRF guards, it would be best to set a wildcard pattern for the origin / referrer, e.g. `https?://[^:/]+(:\d+)?/.*`
* The nginx config is extremely simplistic and would cause issues with AOS-based WebDAV connections to the Alfresco Repository.
* The `/etc/hosts` entries are only relevant for the client access and do not have any impact on the Docker-internal routing.
* Access to load-balanced Share relies on session state. Either nginx must be set to use session-sticky routing or [aldica-provided session replication must be enabled for the Share web application](https://github.com/aldica/aldica/blob/master/docs/Configuration-ShareReference.md#web-session-cache).

**Comment by andreaskring** [comment]
@AFaust Thanks for the feedback. A few comments to your bullets:
1. Good idea - I have implemented this.
1. I tried to do this with the CSRF environment variables discussed earlier, but with no success, so I will go with the solution in 1)
1. Browsing the URLs below `/alfresco/aos` is working fine, but we can tweek the LB if this is needed for more advanced use cases.
1. Yes, I agree. I find it convenient, however, to perform this configuration on my local host, since it makes is easier to "simulate" an environment with real DNS entries. E.g. I found that in some situations when working with Keycloak and Docker-compose everyone must agree on the names.
1. That makes sense - this was the missing peice. I will go for the low hanging fruits for now and have simply added [this](https://github.com/aldica/aldica/blob/feature/multiple-nodes-docker-compose/docker/nginx.conf#L14) to the LB configuration. We can do some more sophisticated aldica/Share configurations later on.

I will make a PR with the changes soon.
